### ISSUE TITLE
Change in the way ObjectMap works? 2.1.1 vs 2.3.1

### ISSUE THREAD

**Issue by ramcaat** [oeg-upm/morph-kgc#119]
Hello Julian --

We are using morph 2.1.1  previously and when we upgraded to 2.3.1 we are getting "**key error "OBJECT_MAP**".. Below is the **json example**
[
  {
    "Table1:Column1": "1111",
    "Table1:Column2": "2222",
    "Table1:Column3": "name",
    "Table1:Column4": "title",
    "Table1:Column5": "No",
    "Table1:Column6": "No",
    "Table1:Column7": "Yes",
    "Table1:Column8": "Bboy",
    "Table1:Column9": "Plain"
  },
  {
    "Table1:Column1": "2222",
    "Table1:Column2": "3333",
    "Table1:Column3": "name1",
    "Table1:Column4": "title1",
    "Table1:Column5": "No1",
    "Table1:Column6": "No1",
    "Table1:Column7": "Yes1",
    "Table1:Column8": "Bboy1",
    "Table1:Column9": "Plain1"
  },
  {
    "Table1:Column1": "3333",
    "Table1:Column2": "4444",
    "Table1:Column3": "name2",
    "Table1:Column4": "title2",
    "Table1:Column5": "No2",
    "Table1:Column6": "No2",
    "Table1:Column7": "Yes2",
    "Table1:Column8": "Bboy2",
    "Table1:Column9": "Plain"
  }
]
**Mapping File :**

@prefix rr: http://www.w3.org/ns/r2rml# .
@prefix rml: http://semweb.mmlab.be/ns/rml# .
@prefix test: https://test.net/test/ .
@prefix rdf: http://www.w3.org/1999/02/22-rdf-syntax-ns#.
@prefix ql: http://semweb.mmlab.be/ns/ql# .

<#ADDRESSLOGICALSOURCE>
rml:iterator "$.*"; 

rml:referenceFormulation ql:JSONPath .

<#LocationMapping>
    rml:logicalSource <#ADDRESSLOGICALSOURCE>;
    a rr:TriplesMap;
    rr:subjectMap [
        rr:template https://test.net/test/{Table1:Column1};
        rr:class test:testClass
    ];
    rr:predicateObjectMap [
        rr:predicate test:Column2;
        rr:objectMap [
             rml:reference "Table1:Column2"
         ];
    ];

.

Please let me know  what has changed compared to our mapping file?

Thanks
Ramu

**Comment by ArenasGuerreroJulian** [comment]
Hi @ramcaat 

is it maybe because you are using a `Table1:Column2` instead `Table1.Column2`?

**Comment by ramcaat** [comment]
The column label is "Table1:column2" Also I have tried with only "column2" and throws same error. Are you able to repro it?

**Comment by ramcaat** [comment]
here it is

Traceback (most recent call last):
  File "test.py", line 23, in <module>
    graph = morph_kgc.materialize(config)
  File "\.venv\lib\site-packages\morph_kgc\__init__.py", line 64, in materialize
    triples = materialize_set(config)
  File "\.venv\lib\site-packages\morph_kgc\__init__.py", line 39, in materialize_set
    mappings_df = retrieve_mappings(config)
  File "\.venv\lib\site-packages\morph_kgc\mapping\mapping_parser.py", line 27, in retrieve_mappings
    mappings = mappings_parser.parse_mappings()
  File "\.venv\lib\site-packages\morph_kgc\mapping\mapping_parser.py", line 395, in parse_mappings
    self._get_from_r2_rml()
  File "\.venv\lib\site-packages\morph_kgc\mapping\mapping_parser.py", line 425, in _get_from_r2_rml
    data_source_mappings_df = self._parse_data_source_mapping_files(section_name)
  File "\.venv\lib\site-packages\morph_kgc\mapping\mapping_parser.py", line 475, in _parse_data_source_mapping_files
    return _transform_mappings_into_dataframe(mapping_graph, section_name)
  File "\.venv\lib\site-packages\morph_kgc\mapping\mapping_parser.py", line 291, in _transform_mappings_into_dataframe
    source_mappings_df['object_join_conditions'] = source_mappings_df['object_map'].map(join_conditions_dict)
  File "\.venv\lib\site-packages\pandas\core\frame.py", line 3505, in __getitem__
    indexer = self.columns.get_loc(key)
  File "\.venv\lib\site-packages\pandas\core\indexes\base.py", line 3623, in get_loc
    raise KeyError(key) from err
KeyError: 'object_map'

**Comment by ArenasGuerreroJulian** [comment]
The mapping is completely bad formmated, if you provide proper data and mappings then I can reproduce it.

From the error message it looks like a problem with a referencing object map

**Comment by ramcaat** [comment]
I am not sure if you meant bad format in the comments or the mapping file .. I can send the file but it will be same content as my first comment.. please use that mapping file and the json or let me know.

**Comment by ArenasGuerreroJulian** [comment]
In the mapping, it is failing to load in RDFlib

**Comment by ramcaat** [comment]
Please try below. Thanks for your patience

**Mapping File**

@prefix rr: <http://www.w3.org/ns/r2rml#> .
@prefix rml: <http://semweb.mmlab.be/ns/rml#> .
@prefix test: <https://test.net/test/>.
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>.
@prefix ql: <http://semweb.mmlab.be/ns/ql#> .
<#ADDRESSLOGICALSOURCE>
rml:iterator "$.*";

rml:referenceFormulation ql:JSONPath .

<#LocationMapping>
    rml:logicalSource <#ADDRESSLOGICALSOURCE>;
    a rr:TriplesMap;
    rr:subjectMap [
        rr:template "https://test.net/{Table1:Column1}";
        rr:class test:Application
    ];

    rr:predicateObjectMap [
        rr:predicate test:applicationId;
        rr:objectMap [
             rml:reference "Table1:Column2"
         ];
    ];
.

**JSON**

[
    {
        "Table1:Column1": "1111",
        "Table1:Column2": "2222",
        "Table1:Column3": "name",
        "Table1:Column4": "title",
        "Table1:Column5": "No",
        "Table1:Column6": "No",
        "Table1:Column7": "Yes",
        "Table1:Column8": "Bboy",
        "Table1:Column9": "Plain"
    },
    {
        "Table1:Column1": "2222",
        "Table1:Column2": "3333",
        "Table1:Column3": "name1",
        "Table1:Column4": "title1",
        "Table1:Column5": "No1",
        "Table1:Column6": "No1",
        "Table1:Column7": "Yes1",
        "Table1:Column8": "Bboy1",
        "Table1:Column9": "Plain1"
    },
    {
        "Table1:Column1": "3333",
        "Table1:Column2": "4444",
        "Table1:Column3": "name2",
        "Table1:Column4": "title2",
        "Table1:Column5": "No2",
        "Table1:Column6": "No2",
        "Table1:Column7": "Yes2",
        "Table1:Column8": "Bboy2",
        "Table1:Column9": "Plain"
    }
]

**Comment by ArenasGuerreroJulian** [comment]
You are missing the `rml:source`. This mapping is working:

```ttl
@prefix  rr: <http://www.w3.org/ns/r2rml#> .
@prefix rml: <http://semweb.mmlab.be/ns/rml#> .
@prefix test: <https://test.net/test/>.
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>.
@prefix ql: <http://semweb.mmlab.be/ns/ql#> .


<#LocationMapping>
    rml:logicalSource [
        rml:source "data.json";
        rml:iterator "$.*";
    ] ;
    rr:subjectMap [
        rr:template "https://test.net/{Table1:Column1}";
        rr:class test:Application
    ];

    rr:predicateObjectMap [
        rr:predicate test:applicationId;
        rr:objectMap [
             rml:reference "Table1:Column2"
         ];
    ];
.
```

**Comment by ramcaat** [comment]
Hi julian --

We pass the input file through config and not directly mention as rml:source. This helps in making our code dynamic.. wonder why this was changed..

```
config = """[DataSourceJSON]
            mappings={} 
            file_path={}
                """.format(new_mapping_file, input_file)
graph = morph_kgc.materialize(config)
```

**Comment by ArenasGuerreroJulian** [comment]
I found the error it is not related to the object map, but with the parsing logical source. #84 removed the OPTIONAL when parsing logical sources, but the OPTIONAL is needed because the source can be specified with `file_path`.

It is now fixed, until next release you can go back to v2.1.1 ot you can grab the latest source from github:

```
pip install git+https://github.com/oeg-upm/morph-kgc.git
```

**Comment by ramcaat** [comment]
I will use 2.3.1 and have blank rml:source for now as we want the latest package for making the empty array issue work. Thanks for your contribution
